vizro-core/tests/unit/vizro/models/_action/test_action.py

```diff
@@ -0,0 +1,205 @@
+        assert callback_inputs == {
+            **{
+                f'{input.split(".")[0]}_{input.split(".")[1]}': dash.State(input.split(".")[0], input.split(".")[1])
+                for input in inputs
+            },
+            "trigger": dash.Input({"action_name": action.id, "type": "action_trigger"}, "data"),
+        }
+        assert callback_outputs == {
+            **{
+                f'{output.split(".")[0]}_{output.split(".")[1]}': dash.Output(
+                    output.split(".")[0], output.split(".")[1], allow_duplicate=True
+                )
+                for output in outputs
+            },
+            "action_finished": dash.Output("action_finished", "data"),
+        }
+        assert action_components == []
+
+    @pytest.mark.parametrize(
+        "custom_action_function, callback_context_outputs_grouping, expected_function_return_value",
+        [
+            (None, [], {"action_finished": None}),
+            (
+                {"component_1_property": "new_value"},
+                ["component_1_property"],
+                {"action_finished": None, "component_1_property": "new_value"},
+            ),
+            ("new_value", ["component_1_property"], {"action_finished": None, "component_1_property": "new_value"}),
+            (
+                ["new_value", "new_value_2"],
+                ["component_1_property", "component_2_property"],
+                {"action_finished": None, "component_1_property": "new_value", "component_2_property": "new_value_2"},
+            ),
+        ],
+        indirect=True,
+    )
+    def test_action_callback_function_return_value_valid(
+        self, custom_action_function, callback_context_outputs_grouping, expected_function_return_value
+    ):
+        action = Action(function=custom_action_function())
+        result = action._action_callback_function()
+        assert result == expected_function_return_value
+
+    @pytest.mark.parametrize(
+        "custom_action_function, callback_context_outputs_grouping",
+        [
+            (None, ["component_1_property"]),
+            ("new_value", []),
+            ("new_value", ["component_1_property", "component_2_property"]),
+            (["new_value"], []),
+            (["new_value"], ["component_1_property", "component_2_property"]),
+            (["new_value", "new_value_2"], ["component_1_property"]),
+            ({"component_1_property": "new_value"}, []),
+            ({"component_1_property": "new_value"}, ["component_1_property", "component_2_property"]),
+            ({"component_1_property": "new_value", "component_2_property": "new_value_2"}, ["component_1_property"]),
+        ],
+        indirect=True,
+    )
+    def test_action_callback_function_return_value_invalid(
+        self, custom_action_function, callback_context_outputs_grouping
+    ):
+        action = Action(function=custom_action_function())
+        with pytest.raises(ValueError):
```

Comment:
Can we be more specific about the correct error message, and potentially annotate the above test cases as to what they represent. I would prefer not to have a lot of cases that are ultimately never imaginable